Add blog about CSI Inline Volumes GA (#35511)
kubernetes/website: content/en/blog/_posts/2022-08-29-csi-inline-volumes-ga.md

- The CSI driver requires `volumeAttributes` that should be restricted to an administrator

## How to use this feature

In order to use this feature, the `CSIDriver` spec must explicitly list `Ephemeral` as one of the supported `volumeLifecycleModes`. Here is a simple example from the [CSI host-path driver](https://github.com/kubernetes-csi/csi-driver-host-path).

```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: hostpath.csi.k8s.io
spec:
  volumeLifecycleModes:
  - Persistent
  - Ephemeral
  podInfoOnMount: true
  fsGroupPolicy: File
```

Any pod spec may then reference that CSI driver to create an inline volume, as in this example.

```
kind: Pod
apiVersion: v1
metadata:
  name: my-csi-app-inline
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: topology.hostpath.csi/node
            operator: Exists
  containers:
    - name: my-frontend
      image: busybox
      volumeMounts:
      - mountPath: "/data"
        name: my-csi-volume
      command: [ "sleep", "1000000" ]
  volumes:
    - name: my-csi-volume
      csi:

>         driver: hostpath.csi.k8s.io

Reviewer: Host path is kind of a bad example IMO. doesn't show anything it can do beyond what k8s natively can do.

Reviewer 2: I'd like to pick an in-project CSI driver if we can, but I agree: `hostPath` / `hostpath.csi.k8s.io` is a bit boring.

How about Secrets Store?

Reviewer: image populator is still in project. secretstore would be a good one too though.

Author: Thanks for the feedback, I updated the example to use the secret store csi driver.

```

> If the driver supports any volume attributes, these may also be provided as part of the pod spec.

Reviewer 2: ```suggestion
If the driver supports any volume attributes, you can provider these as part of the `spec` for the Pod as well:
```

```
      csi:

>         driver: example.csi.k8s.io

Reviewer 2: Is this valid:
```suggestion
        driver: block.csi.vendor.example
```
?

If so, let's illustrate that driver IDs don't have to be official Kubernetes project ones.

## Example Use Cases

Two existing CSI drivers that support the `Ephemeral` volume lifecycle mode are the Secrets Store CSI Driver and the Cert-Manager CSI Driver.

The [Secrets Store CSI Driver](https://github.com/kubernetes-sigs/secrets-store-csi-driver) allows users to mount secrets from external secret stores into a pod as an inline volume. This can be useful when the secrets are stored in an external managed service or Vault instance.

The [Cert-Manager CSI Driver](https://github.com/cert-manager/csi-driver) works along with [cert-manager](https://cert-manager.io/) to seamlessly request and mount certificate key pairs into a pod. This allows the certificates to be renewed and updated in the application pod automatically.

## Security Considerations

Special consideration should be given to which CSI drivers may be used as inline volumes. `volumeAttributes` are typically controlled through the `StorageClass`, and may contain attributes that should remain restricted to the cluster administrator. Allowing a CSI driver to be used for inline ephmeral volumes means that any user with permission to create pods may also provide `volumeAttributes` to the driver through a pod spec.

> Cluster administrators may choose to remove `Ephemeral` from `volumeLifecycleModes` in the CSIDriver spec to prevent the driver from being used as an inline ephemeral volume, or use an [admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) to restrict how the driver is used.

Reviewer 2: ```suggestion
Cluster administrators may choose to omit (or remove) `Ephemeral` from `volumeLifecycleModes` in the CSIDriver spec to prevent the driver from being used as an inline ephemeral volume, or use an [admission webhook](/docs/reference/access-authn-authz/extensible-admission-controllers/) to restrict how the driver is used.
```

## What's new in 1.25?

There are a couple of new bug fixes related to this feature in 1.25, and the [CSIInlineVolume feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) has been locked to `True` with the graduation to GA. There are no new API changes, so users of this feature during beta should not notice any significant changes aside from these bug fixes.

- [#89290 - CSI inline volumes should support fsGroup](https://github.com/kubernetes/kubernetes/issues/89290)
- [#79980 - CSI volume reconstruction does not work for ephemeral volumes](https://github.com/kubernetes/kubernetes/issues/79980)

## When to use this feature

CSI inline volumes are meant for simple local volumes that should follow the lifecycle of the pod. They may be useful for providing secrets, configuration data, or other special-purpose storage to the pod from a CSI driver.

A CSI driver is not suitable for inline use when:
- The volume needs to persist longer than the lifecycle of a pod
- Volume snapshots, cloning, or volume expansion are required
- The CSI driver requires `volumeAttributes` that should be restricted to an administrator

> ## How to use this feature

Reviewer 3: Add a link to secrete store CSI driver as an example implementation?  https://github.com/kubernetes-sigs/secrets-store-csi-driver#features

Author: Good idea, I added a new section with example use cases.

Reviewer: A couple of examples:
Using a container image to ship data rather then programs: https://github.com/kubernetes-csi/csi-driver-image-populator/blob/master/examples/pod.yaml
Support safe privileged/unprivilaged communication https://github.com/spiffe/spiffe-csi/blob/main/example/config/workload.yaml
